### PR TITLE
Change prologue details to prepare for fsdp as a transform

### PR DESCRIPTION
**Pull request by t-vi**
- support tracing a model on meta device (to enable tracing before materializing / sharding)
- use model.get_parameter / model.get_buffer to acquire parameters and buffers to more easily divert these.

### CODE REVIEW

File: thunder/core/jit_ext.py
Comment (IvanYashchuk) [inline thunder/core/jit_ext.py:1]: I must confess I don't understand what's going on in this file and I'm glad I don't need to.

File: thunder/tests/test_jit_general.py [hunk 1]
```diff
@@ -629,7 +629,7 @@ def test_nanogpt():
 )
 @pytest.mark.parametrize(
     "device",
-    ("cpu", "cuda"),
+    ("cpu", "cuda", "meta"),
```
Comment (IvanYashchuk) [inline thunder/tests/test_jit_general.py:632]: Do we call PyTorch's meta functions anywhere in the pipeline? What's the state of meta tensors in PyTorch?

**Review by IvanYashchuk — APPROVED** [review]
How do prologue traces change with this PR?

I don't see yet how this will be connected with the "fsdp as a transform" work and I'm very curious to see where this leads to!

File: thunder/core/devices.py [hunk 2]
```diff
@@ -14,13 +14,15 @@
...
     CUDA = auto()
+    META = auto()
 
 
 all_devicetypes = (DeviceType.CPU, DeviceType.CUDA)
 
 _devicetype_prettyprint_map = {
     DeviceType.CPU: "cpu",
     DeviceType.CUDA: "cuda",
+    DeviceType.META: "meta",
```
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/core/devices.py:25]: I'm not attached to it any particular way, this is inspired by the thunder.distributed.fsdp 's use of meta.

**Comment by t-vi** [comment]
Thank you @IvanYashchuk @crcrpar . Let's think about the meta bits separately. I'm happy to switch to whatever works best for fsdp.

**Pull request merged**